webapp/src/snippetBuilder.tsx

```diff
@@ -0,0 +1,398 @@
+        };
+
+        this.hide = this.hide.bind(this);
+        this.cancel = this.cancel.bind(this);
+        this.confirm = this.confirm.bind(this);
+        this.backPage = this.backPage.bind(this);
+        this.nextPage = this.nextPage.bind(this);
+    }
+
+    /**
+     * Creates a hashmap with answerToken keys and the default value pair as 
+     * provided by our config file.
+     */
+    buildDefaults() {
+        const { config } = this.state;
+        const defaults: any = {};
+
+        for (const question of config.questions) {
+            const { inputs } = question;
+            for (const input of inputs) {
+                const { defaultAnswer, answerToken } = input;
+                defaults[answerToken] = defaultAnswer;
+            }
+        }
+
+        this.setState({ defaults });
+    }
+
+    /**
+     * Calls build defaults on mount to create the defaults hashmap.
+     */
+    componentDidMount() {
+        // Sets default values
+        this.buildDefaults();
+    }
+
+    /**
+     * @param output - Takes in a string and returns the tokenized output
+     * Loops over each token previously added to defaults and replaces with the answer value if one
+     * exists. Otherwise it replaces the token with the provided default value.
+     */
+    replaceTokens(tsOutput: string) {
+        const { answers, defaults } = this.state;
+        let tokenizedOutput = tsOutput;
+        const tokens = Object.keys(defaults);
+
+        // Replaces output tokens with answer if available or default value
+        for (let token of tokens) {
+            if (answers[token]) {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(answers[token]);
+            }
+            else {
+                tokenizedOutput = tokenizedOutput.split(`$${token}`).join(defaults[token]);
+            }
+        }
+
+        return tokenizedOutput;
+    }
+
+    /**
+     * 
+     * @param output - Accepts an output to convert to markdown
+     * This attaches three backticks to the front followed by an output type (blocks, lang)
+     * The current output is then tokenized and three backticks are appended to the end of the string.
+     */
+    generateOutputMarkdown(tsOutput: string) {
+        const { config } = this.state;
+        // Attaches starting and ending line based on output type
+        let md = `\`\`\`${config.outputType}\n`;
+        md += this.replaceTokens(tsOutput);
+        md += `\n\`\`\``;
+
+        return md
+    }
+
+    /**
+     * Hides the modal
+     */
+    hide() {
+        this.setState({
+            visible: false
+        });
+    }
+
+    /**
+     * 
+     * @param projectView - used to access what would traditionally be in the parent prop
+     * @param mainWorkspace  - used to append the final xml to the DOM
+     */
+    show(projectView: pxt.editor.IProjectView, mainWorkspace: Blockly.Workspace) {
+        pxt.tickEvent('snippetBuilder.show', null, { interactiveConsent: false });
```

Comment:
isn't the sprite builder opened via a button click? in that case, interactiveConsent is true.

Reply:
I added a catch that throws an exception and reports the error.